In [1]:
from pathlib import Path
import pandas as pd
import numpy as np

CURRENT_DIR = Path.cwd()

if CURRENT_DIR.name == "notebooks":
    PROJECT_ROOT = CURRENT_DIR.parent
else:
    PROJECT_ROOT = CURRENT_DIR

DATA_PATH = PROJECT_ROOT / "data" / "raw" / "credit_default_raw.xls"

print("Project root:", PROJECT_ROOT)
print("Dataset path:", DATA_PATH)
print("File exists:", DATA_PATH.exists())

df = pd.read_excel(DATA_PATH, header=1)

print("\nDataset shape:", df.shape)

df.head()

Project root: h:\Credit_Risk_Modeling
Dataset path: h:\Credit_Risk_Modeling\data\raw\credit_default_raw.xls
File exists: True

Dataset shape: (30000, 25)


,ID,LIMIT_BAL,SEX,EDUCATION,MARRIAGE,AGE,PAY_0,PAY_2,PAY_3,PAY_4,...,BILL_AMT4,BILL_AMT5,BILL_AMT6,PAY_AMT1,PAY_AMT2,PAY_AMT3,PAY_AMT4,PAY_AMT5,PAY_AMT6,default payment next month
0,1,20000,2,2,1,24,2,2,-1,-1,...,0,0,0,0,689,0,0,0,0,1
1,2,120000,2,2,2,26,-1,2,0,0,...,3272,3455,3261,0,1000,1000,1000,0,2000,1
2,3,90000,2,2,2,34,0,0,0,0,...,14331,14948,15549,1518,1500,1000,1000,1000,5000,0
3,4,50000,2,2,1,37,0,0,0,0,...,28314,28959,29547,2000,2019,1200,1100,1069,1000,0
4,5,50000,1,2,1,57,-1,0,-1,0,...,20940,19146,19131,2000,36681,10000,9000,689,679,0


In [2]:
# Kiểm tra cấu trúc tổng quan của dataset

print("Số dòng:", df.shape[0])
print("Số cột:", df.shape[1])

print("\nDanh sách cột:")
for i, col in enumerate(df.columns, start=1):
    print(f"{i:02d}. {col}")

print("\nKiểu dữ liệu:")
df.info()

Số dòng: 30000
Số cột: 25

Danh sách cột:
01. ID
02. LIMIT_BAL
03. SEX
04. EDUCATION
05. MARRIAGE
06. AGE
07. PAY_0
08. PAY_2
09. PAY_3
10. PAY_4
11. PAY_5
12. PAY_6
13. BILL_AMT1
14. BILL_AMT2
15. BILL_AMT3
16. BILL_AMT4
17. BILL_AMT5
18. BILL_AMT6
19. PAY_AMT1
20. PAY_AMT2
21. PAY_AMT3
22. PAY_AMT4
23. PAY_AMT5
24. PAY_AMT6
25. default payment next month

Kiểu dữ liệu:
<class 'pandas.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 25 columns):
 #   Column                      Non-Null Count  Dtype
---  ------                      --------------  -----
 0   ID                          30000 non-null  int64
 1   LIMIT_BAL                   30000 non-null  int64
 2   SEX                         30000 non-null  int64
 3   EDUCATION                   30000 non-null  int64
 4   MARRIAGE                    30000 non-null  int64
 5   AGE                         30000 non-null  int64
 6   PAY_0                       30000 non-null  int64
 7   PAY_2                       

In [3]:
target_col = "default payment next month"

target_summary = pd.DataFrame({
    "Count": df[target_col].value_counts(),
    "Percent": (df[target_col].value_counts(normalize=True) * 100).round(2)
})

target_summary

,Count,Percent
default payment next month,,
0,23364,77.88
1,6636,22.12


In [4]:
duplicate_count = df.duplicated().sum()

print("Số dòng bị duplicate:", duplicate_count)
print("Tỷ lệ duplicate:", round(duplicate_count / len(df) * 100, 2), "%")

Số dòng bị duplicate: 0
Tỷ lệ duplicate: 0.0 %


In [5]:
print("Số ID duy nhất:", df["ID"].nunique())
print("Tổng số dòng:", len(df))
print("Số ID bị trùng:", df["ID"].duplicated().sum())

Số ID duy nhất: 30000
Tổng số dòng: 30000
Số ID bị trùng: 0


In [6]:
categorical_cols = ["SEX", "EDUCATION", "MARRIAGE"]

for col in categorical_cols:
    print(f"\n===== {col} =====")
    print(df[col].value_counts().sort_index())


===== SEX =====
SEX
1    11888
2    18112
Name: count, dtype: int64

===== EDUCATION =====
EDUCATION
0       14
1    10585
2    14030
3     4917
4      123
5      280
6       51
Name: count, dtype: int64

===== MARRIAGE =====
MARRIAGE
0       54
1    13659
2    15964
3      323
Name: count, dtype: int64


In [7]:
payment_status_cols = [
    "PAY_0",
    "PAY_2",
    "PAY_3",
    "PAY_4",
    "PAY_5",
    "PAY_6"
]

for col in payment_status_cols:
    print(f"\n===== {col} =====")
    print(df[col].value_counts().sort_index())


===== PAY_0 =====
PAY_0
-2     2759
-1     5686
 0    14737
 1     3688
 2     2667
 3      322
 4       76
 5       26
 6       11
 7        9
 8       19
Name: count, dtype: int64

===== PAY_2 =====
PAY_2
-2     3782
-1     6050
 0    15730
 1       28
 2     3927
 3      326
 4       99
 5       25
 6       12
 7       20
 8        1
Name: count, dtype: int64

===== PAY_3 =====
PAY_3
-2     4085
-1     5938
 0    15764
 1        4
 2     3819
 3      240
 4       76
 5       21
 6       23
 7       27
 8        3
Name: count, dtype: int64

===== PAY_4 =====
PAY_4
-2     4348
-1     5687
 0    16455
 1        2
 2     3159
 3      180
 4       69
 5       35
 6        5
 7       58
 8        2
Name: count, dtype: int64

===== PAY_5 =====
PAY_5
-2     4546
-1     5539
 0    16947
 2     2626
 3      178
 4       84
 5       17
 6        4
 7       58
 8        1
Name: count, dtype: int64

===== PAY_6 =====
PAY_6
-2     4895
-1     5740
 0    16286
 2     2766
 3      184
 4       49


In [8]:
df[payment_status_cols].agg(["min", "max"])

,PAY_0,PAY_2,PAY_3,PAY_4,PAY_5,PAY_6
min,-2,-2,-2,-2,-2,-2
max,8,8,8,8,8,8


In [9]:
for col in payment_status_cols:
    summary = (
        df.groupby(col)[target_col]
        .agg(
            Customers="count",
            Defaults="sum",
            Default_Rate="mean"
        )
    )

    summary["Default_Rate"] = (
        summary["Default_Rate"] * 100
    ).round(2)

    print(f"\n===== {col} =====")
    print(summary)


===== PAY_0 =====
       Customers  Defaults  Default_Rate
PAY_0                                   
-2          2759       365         13.23
-1          5686       954         16.78
 0         14737      1888         12.81
 1          3688      1252         33.95
 2          2667      1844         69.14
 3           322       244         75.78
 4            76        52         68.42
 5            26        13         50.00
 6            11         6         54.55
 7             9         7         77.78
 8            19        11         57.89

===== PAY_2 =====
       Customers  Defaults  Default_Rate
PAY_2                                   
-2          3782       691         18.27
-1          6050       966         15.97
 0         15730      2503         15.91
 1            28         5         17.86
 2          3927      2184         55.61
 3           326       201         61.66
 4            99        50         50.51
 5            25        15         60.00
 6            12   

In [10]:
numeric_check_cols = [
    "LIMIT_BAL",
    "AGE",
    "BILL_AMT1",
    "BILL_AMT2",
    "BILL_AMT3",
    "BILL_AMT4",
    "BILL_AMT5",
    "BILL_AMT6",
    "PAY_AMT1",
    "PAY_AMT2",
    "PAY_AMT3",
    "PAY_AMT4",
    "PAY_AMT5",
    "PAY_AMT6"
]

df[numeric_check_cols].describe().T[
    ["min", "mean", "50%", "max"]
].round(2)

,min,mean,50%,max
LIMIT_BAL,10000.0,167484.32,140000.0,1000000.0
AGE,21.0,35.49,34.0,79.0
BILL_AMT1,-165580.0,51223.33,22381.5,964511.0
BILL_AMT2,-69777.0,49179.08,21200.0,983931.0
BILL_AMT3,-157264.0,47013.15,20088.5,1664089.0
BILL_AMT4,-170000.0,43262.95,19052.0,891586.0
BILL_AMT5,-81334.0,40311.40,18104.5,927171.0
BILL_AMT6,-339603.0,38871.76,17071.0,961664.0
PAY_AMT1,0.0,5663.58,2100.0,873552.0
PAY_AMT2,0.0,5921.16,2009.0,1684259.0


## Data Understanding Summary

- Dataset contains 30,000 customers and 25 variables.
- No missing values were detected.
- No full-row duplicates were detected.
- Target variable:
  - Non-default: 23,364 customers (77.88%)
  - Default: 6,636 customers (22.12%)
- The target is moderately imbalanced, so accuracy alone will not be sufficient for model evaluation.
- SEX contains valid categories only.
- EDUCATION contains undocumented categories 0, 5, and 6.
- MARRIAGE contains undocumented category 0.
- Payment status variables PAY_0 to PAY_6 range from -2 to 8.
- Payment delay history shows a strong relationship with default risk.
- Negative BILL_AMT values will not be removed without further investigation.
- Large payment and bill values will be examined during EDA before any outlier treatment.   